### ISSUE TITLE
Use withCallingHandlers + on.exit instead of tryCatch

### ISSUE THREAD

**Issue by hadley** [r-lib/crancache#16]
This approach looks more correct to me:

```R
  if (should_update_crancache()) {
    add_built_binaries <- should_add_binaries()

    on.exit(update_cache(
      destdir, binaries = add_built_binaries, warnings, errors, lib,
      timestamp, args
    ))
  }

  withCallingHandlers(
    utils::install.packages(
      pkgs = pkgs,
      lib = lib,
      repos = myrepos,
      ## We don't specify contriburl, on purpose
      method = method,
      available = NULL,                   # overwritten
      destdir = destdir,
      dependencies = dependencies,
      type = type,
      ...),
    warning = function(w) { warnings <- append(warnings, w); warning(w) }
  )
```

`tryCatch()` installs exiting handlers, so the first warning will terminate execution of `install.packages()`.

**Comment by hadley** [comment]
This fixes the problem I was having, although I'm not sure why. Presumably something important was getting swallowed.

**Comment by gaborcsardi** [comment]
Btw. then we probably also want to change the rest of the functions, i.e. `download_packages`, `update_packages`. Feel free to do that, or I can do it as well.

**Comment by gaborcsardi** [comment]
I vaguely remember something about `on.exit` not being appropriate for this, but I just cannot recall it now.....  so I'll commit these changes now, we'll see if there is anything wrong with `on.exit`.....

**Comment by gaborcsardi** [comment]
Closed by 07dd046c9968a54d4f14338d7f997d6c6d7e0792
